Add support for specifying required recipient contact fields for Disbursement requests

codeOfRobin opened the pull request:
#### 0feced788ade150bbdb6a80e0934add69b4a30dd
<pre>
Add support for specifying required recipient contact fields for Disbursement requests

<a href="https://bugs.webkit.org/show_bug.cgi?id=272526">https://bugs.webkit.org/show_bug.cgi?id=272526</a>
<a href="https://rdar.apple.com/115776112">rdar://115776112</a>
Reviewed by Abrar Rahman Protyasha.

We add support for the requisite types to specify recipient contact information in a DisbursementRequest

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
* Source/WebCore/Modules/applepay/ApplePayContactField.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayContactField.h:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/ApplePay/DisbursementPaymentRequest.h:
* Source/WebKit/Shared/ApplePay/cocoa/DisbursementPaymentRequestCocoa.mm:
(WebKit::platformDisbursementPaymentRequest):
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/277953@main">https://commits.webkit.org/277953@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08acaadb8d3cba7c3b50db44c173282d364d206c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28213 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45070 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51305 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40058 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42248 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23317 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43418 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7055 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43923 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53598 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24051 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47368 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25314 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46346 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10791 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26123 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25034 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -1,116 +1,187 @@
+        test((test) => {
+            const method = validPaymentMethod();
+            method.data.features = ["paymentRequestDisbursements"];
+            new PaymentRequest([method], validPaymentDetails());
```
I wonder if we should assure this with something like:

```suggestion
            try {
               method.data.features = ["not a thing"];
               new PaymentRequest([method], validPaymentDetails());
               assert_unreached("Should have thrown")
            } catch {
               method.data.features = ["paymentRequestDisbursements"];
               new PaymentRequest([method], validPaymentDetails());              
            }
            
```

codeOfRobin replied:
Not opposed to it - done!

marcoscaceres reviewed (approved):
Nit that there is still a couple of whitespace issues on the test file, but otherwise approving on the JS tests (non-reviewer ✅)

aprotyas commented on Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
```diff
@@ -52,7 +52,8 @@
     if (webDisbursementPaymentRequest) {
-        auto disbursementRequest = platformDisbursementPaymentRequest(request, originatingURL);
+        auto disbursementRequest = platformDisbursementPaymentRequest(request, originatingURL, webDisbursementPaymentRequest->requiredRecipientContactFields);
+        // TODO: get webPageProxyID etc added
```
This is no longer relevant, so you can drop this comment!

codeOfRobin replied:
argh bad rebase, sorry about that!

aprotyas commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -1228,6 +1232,13 @@ enum class WebCore::ApplePayErrorCode : uint8_t {
+    std::optional<Vector<WebCore::ApplePayContactField>> requiredRecipientContactFields
+};
+#endif
+
```
Nit: two new lines here, so one can go.

aprotyas reviewed (approved):
LGTM

aprotyas reviewed (commented):
Actually, the code change is good, but the commit message formatting is off. Can you generate it again with `git webkit`? It should follow this format.

```
<title>
<bugzilla>
<radar>

Reviewed by NOBODY (OOPS!).

<description>

<file-change-list>
```

Commit:
Add support for specifying required recipient contact fields for Disbursement requests

https://bugs.webkit.org/show_bug.cgi?id=272526
rdar://115776112
Reviewed by Abrar Rahman Protyasha.

We add support for the requisite types to specify recipient contact information in a DisbursementRequest

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https-expected.txt:
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
* Source/WebCore/Modules/applepay/ApplePayContactField.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayContactField.h:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/ApplePay/DisbursementPaymentRequest.h:
* Source/WebKit/Shared/ApplePay/cocoa/DisbursementPaymentRequestCocoa.mm:
(WebKit::platformDisbursementPaymentRequest):
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/277953@main
